File: articles/biostatistics-experimental-design-biology/python/permutation_test.py
"""
Permutation test workflow.

Run:
    python python/permutation_test.py
"""

from pathlib import Path

import numpy as np
import pandas as pd


ARTICLE_DIR = Path(__file__).resolve().parents[1]
DATA_PATH = ARTICLE_DIR / "data" / "two_group_measurements.csv"


def main() -> None:
    rng = np.random.default_rng(42)

    data = pd.read_csv(DATA_PATH)

    control = data.loc[data["group"] == "control", "value"].to_numpy()
    treated = data.loc[data["group"] == "treated", "value"].to_numpy()

    observed_difference = treated.mean() - control.mean()
    combined = np.concatenate([control, treated])
    n_control = len(control)

    n_permutations = 10000
    null = np.empty(n_permutations)

    for i in range(n_permutations):
        shuffled = rng.permutation(combined)
        null[i] = shuffled[n_control:].mean() - shuffled[:n_control].mean()

    p_value = np.mean(np.abs(null) >= abs(observed_difference))

    result = pd.DataFrame(
        {
            "observed_difference": [observed_difference],
            "permutation_p_value": [p_value],
            "null_mean": [null.mean()],
            "null_sd": [null.std(ddof=1)],
        }
    )

    print(result.round(6).to_string(index=False))


if __name__ == "__main__":
    main()

File: articles/biostatistics-experimental-design-biology/r/permutation_test.R
# Permutation test workflow in R.

data_path <- file.path("data", "two_group_measurements.csv")
if (!file.exists(data_path)) {
  data_path <- file.path("..", "data", "two_group_measurements.csv")
}

set.seed(42)

data <- read.csv(data_path)

control <- data$value[data$group == "control"]
treated <- data$value[data$group == "treated"]

observed_difference <- mean(treated) - mean(control)
combined <- c(control, treated)
n_control <- length(control)

n_permutations <- 10000

null <- replicate(n_permutations, {
  shuffled <- sample(combined)
  mean(shuffled[(n_control + 1):length(shuffled)]) - mean(shuffled[1:n_control])
})

p_value <- mean(abs(null) >= abs(observed_difference))

summary_df <- data.frame(
  observed_difference = observed_difference,
  permutation_p_value = p_value,
  null_mean = mean(null),
  null_sd = sd(null)
)

print(round(summary_df, 6))
